pkg/controllers/node/hostendpoints.go

```diff
@@ -0,0 +1,296 @@
+				continue
+			}
+		}
+
+		log.Info("successfully synced all hostendpoints")
+		return nil
+	}
+	return fmt.Errorf("too many retries when syncing all hostendpoints")
+}
+
+// syncAutoHostendpoint syncs the auto hostendpoint for the given node.
+func (c *NodeController) syncAutoHostendpoint(node *api.Node) error {
+	hepName := c.generateAutoHostendpointName(node.Name)
+	log.Debugf("syncing hostendpoint %q from node %+v", hepName, node)
+
+	// Try getting the host endpoint.
+	expectedHep := c.generateAutoHostendpointFromNode(node)
+	currentHep, err := c.calicoClient.HostEndpoints().Get(c.ctx, hepName, options.GetOptions{})
+	if err != nil {
+		switch err.(type) {
+		case errors.ErrorResourceDoesNotExist:
+			log.Infof("host endpoint %q doesn't exist, creating it", node.Name)
+			if _, err := c.createAutoHostendpoint(node); err != nil {
+				log.WithError(err).Warnf("failed to create host endpoint %q, retrying", node.Name)
+				return err
+			}
+		default:
+			log.WithError(err).Warnf("failed to get host endpoint %q, retrying", node.Name)
+			return err
+		}
+	} else if err := c.updateHostendpoint(currentHep, expectedHep); err != nil {
+		log.WithError(err).Warnf("failed to update hostendpoint %q, retrying", currentHep.Name)
+		return err
+	}
+
+	log.WithField("hep.Name", expectedHep.Name).Info("successfully synced hostendpoint")
+	return nil
+}
+
+// syncAutoHostendpointWithRetries syncs the auto hostendpoint for the given
+// node, retrying a few times if needed.
+func (c *NodeController) syncAutoHostendpointWithRetries(node *api.Node) error {
+	for n := 1; n <= 5; n++ {
+		log.Debugf("syncing hostendpoint for node %q. attempt #%v", node.Name, n)
+		if err := c.syncAutoHostendpoint(node); err != nil {
+			log.WithError(err).Infof("failed to sync host endpoint for node %q, retrying", node.Name)
+			time.Sleep(retrySleepTime)
+			continue
+		}
+		return nil
+	}
+	return fmt.Errorf("too many retries when syncing hostendpoint for node %q", node.Name)
+}
+
+// listAutoHostendpoints returns a map of auto hostendpoints keyed by the
+// hostendpoint's name.
+func (c *NodeController) listAutoHostendpoints() (map[string]api.HostEndpoint, error) {
+	time.Sleep(c.rl.When(RateLimitCalicoList))
+	heps, err := c.calicoClient.HostEndpoints().List(c.ctx, options.ListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("could not list hostendpoints: %v", err.Error())
+	}
+	c.rl.Forget(RateLimitCalicoList)
+	m := make(map[string]api.HostEndpoint)
+	for _, h := range heps.Items {
+		if isAutoHostendpoint(&h) {
+			m[h.Name] = h
+		}
+	}
+	return m, nil
+}
+
+// deleteHostendpoint removes the specified hostendpoint, optionally retrying
+// the operation a few times until it succeeds.
+func (c *NodeController) deleteHostendpoint(hepName string) error {
+	log.Debugf("deleting hostendpoint %q", hepName)
+	time.Sleep(c.rl.When(RateLimitCalicoDelete))
+	_, err := c.calicoClient.HostEndpoints().Delete(c.ctx, hepName, options.DeleteOptions{})
+	if err != nil {
```

Comment:
I think we should handle "DoesNotExist" errors here. If we try to delete it and it doesn't exist, that's ok :)

Reply:
7ab6868